### ISSUE TITLE
Increasing memory usage of train.py

### ISSUE THREAD

**Issue by muupan** [Russell91/ReInspect#1]
First of all, thank you for sharing the impressive code!

I ran train.py in its GPU mode on Ubuntu 14.04 with CUDA 7.0 and GTX 970 following the instruction in README.md. Though it successfully trained the network, its usage of the host memory (not the GPU device memory) increased slowly but constantly as the training proceeded, almost using up my 32GB memory eventually. Has anyone experienced the same problem?

The versions of ReInspect and apollocaffe are ~~https://github.com/Russell91/ReInspect/commit/2d9739bcfccb12232a36acaee8c8d0db6e73a2ec~~ https://github.com/Russell91/ReInspect/commit/0803bd2e106e905933e8c49991a400c82094572b and https://github.com/Russell91/apollocaffe/commit/1239e65945b19c55095dbbd307be546e8e97499f respectively.

**Comment by Russell91** [comment]
Looking into this.

**Comment by Russell91** [comment]
Thanks for the bug report. Was able to reproduce and identified the cause of the issue as a memory leak in the hungarian_loss_layer. See https://github.com/Russell91/apollocaffe/commit/037c8edbc8e6e3d74f13beb00e3f6647614dcf02 for the fix.

As some vindication on our end, this bug was only introduced just a few weeks ago when we switched from DLib's hungarian implementation to a very lightweight c implementation to remove the bulky dlib dependency for you guys. Unfortunately, the suggested API for that new library contains the memory leak that you discovered. I've notified the author, but you'll have to grab the latest ApolloCaffe to get the patch.
